### ISSUE TITLE
Should the target of libraries loaded from islands be es2022 by default?

### ISSUE THREAD

**Issue by ayame113** [denoland/fresh#1125]
By default, esm.sh checks the User-Agent header to determine the build target. ([src](https://esm.sh/#docs:~:text=in%20development%20mode.-,ESBuild%20Options,-By%20default%2C%20esm))

So if people import esm.sh from islands, the bundle target will be Deno. (ref: #988)
Maybe a proper User-Agent needs to be specified when bundling with esbuild? Then people shouldn't have to manually specify `?target=es2022` in their dependency URLs.

**Comment by adamgreg** [comment]
I'd be very interested in any solution to this. I'm new to Deno, and trying to import a component in an Island from esm.sh. It seems that even if I add `?target=es2022` to the import URL the browser ends up importing dependencies of the imported module from `deno` URLs instead of `es2022` ones.

EDIT: I got there in the end, but only once I'd added every dependency (and dependency of a dependency, etc.) to my import map using esm.sh URLs with `?target=es2022` and the `*` to make all dependencies external, as I couldn't find a way to make the `target` option propagate through the import graph.

**Comment by adamgreg** [comment]
I'd be very interested in any solution to this. I'm new to Fresh, and trying to use a component library with lots of dependencies in my islands. I eventually got it working, but only once I'd added every dependency (and dependency of a dependency, etc.) to my import map using esm.sh URLs with `?target=es2022` and the `*` to make all dependencies external, as I couldn't find a way to make the target option propagate through the import graph.

**Comment by lucacasonato** [comment]
`npm:` specifiers will mean native NPM bundling rather than using esm.sh for this. Let's wait for that :)

**Comment by ayame113** [comment]
Can this be reconsidered?

For me, it's a bit annoying to ask the library author "sorry, please add `?target=es2020` to the import statements in your library" every time you run into a problem.

- https://github.com/denoland/fresh/pull/1490/files
- https://github.com/denoland/saaskit/pull/357/files
- https://github.com/denoland/deno-gfm/pull/35/files

**Comment by marvinhagemeister** [comment]
Good call. Maybe we can solve this by adding an esbuild plugin that appends the target parameter during resolution automatically.

**Comment by marvinhagemeister** [comment]
Since Fresh 1.5 you can set the esbuild target option in the Fresh configuration via `build.target`. This allows you to down transpile the code for the specific browser requirements of the project.

**Comment by adamgreg** [comment]
Is that definitely going to address the issue? I believe the problem was that esm.sh will serve code including Deno API calls, unless you add `?target=es2022` to the request or use a different `User-Agent` header. Unless those are now replaced later, when the islands are built, I'm not sure this is resolved.

Contrary to my comment in May, I believe esm.sh will now properly propagate the target to transitive dependencies.

**Comment by bartlomieju** [comment]
Closing — no longer relevant. Fresh 2.x uses Vite instead of esbuild for bundling, and esm.sh imports are discouraged (see [troubleshooting docs](https://fresh.deno.dev/docs/advanced/troubleshooting#dont-use-esmsh)). Dependencies should come from npm/JSR via `deno.json` imports, which Vite resolves and bundles correctly for the browser target.
